src/gadgets/sumcheck.rs

```diff
@@ -0,0 +1,336 @@
+  degree_bound: usize,
+  polys: Vec<Vec<AllocatedNum<Scalar>>>,
+  challenges: Vec<AllocatedNum<Scalar>>, // this is while Transcript circuit is not ready
+) -> Result<(AllocatedNum<Scalar>, Vec<AllocatedNum<Scalar>>), SynthesisError>
+where
+  Scalar: PrimeField + PrimeFieldBits,
+  CS: ConstraintSystem<Scalar>,
+{
+  let mut e = claim;
+  for (i, poly) in polys.iter().enumerate() {
+    // verify poly degree // TODO move check to constraints
+    if poly.len() - 1 != degree_bound {
+      panic!(
+        "poly degree {:?} != degree_bound {:?}",
+        poly.len() - 1,
+        degree_bound
+      );
+    }
+
+    // eval at zero
+    let eval_0 = poly[0].clone();
+    // eval at one
+    let eval_1 = eval_at_one(cs.namespace(|| format!("sc_eval_at_one {}", i)), poly)?;
+
+    cs.enforce(
+      || format!("(s(0)+s(1)) * 1 = s(r) {}", i),
+      |lc| lc + eval_0.get_variable() + eval_1.get_variable(),
+      |lc| lc + CS::one(),
+      |lc| lc + e.get_variable(),
+    );
+
+    let r_i = challenges[i].clone(); // this is while Transcript circuit is not ready
+
+    e = uni_evaluate(
+      cs.namespace(|| format!("sc_unieval {}", i)),
+      poly.clone(),
+      r_i,
+    )?;
+  }
+
+  // note: challenges will be replaced by r vector once the Transcript circuit is ready
+  Ok((e, challenges))
+}
+
+pub fn eval_at_one<Scalar, CS>(
+  mut cs: CS,
+  poly: &Vec<AllocatedNum<Scalar>>,
+) -> Result<AllocatedNum<Scalar>, SynthesisError>
+where
+  Scalar: PrimeField + PrimeFieldBits,
+  CS: ConstraintSystem<Scalar>,
+{
+  let mut eval: Num<Scalar> = Num::<Scalar>::from(poly[0].clone());
+  for (i, coeff) in poly.iter().enumerate().skip(1) {
+    // logic: eval_aux = eval + coef
+    let eval_aux = eval.clone().add(&Num::<Scalar>::from(coeff.clone()));
+    // r1cs constr: (eval + coeff) * 1 = eval_aux
+    let allocated_eval: AllocatedNum<Scalar> =
+      AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval.get_value().unwrap()))?;
+    let allocated_eval_aux: AllocatedNum<Scalar> =
+      AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval_aux.get_value().unwrap()))?;
+    cs.enforce(
+      || format!("eval + coeff = eval_aux{}", i),
+      |lc| lc + allocated_eval.get_variable() + coeff.get_variable(),
+      |lc| lc + CS::one(),
+      |lc| lc + allocated_eval_aux.get_variable(),
+    );
+    eval = eval_aux;
+  }
+  let allocated_eval: AllocatedNum<Scalar> =
+    AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval.get_value().unwrap()))?;
+
+  Ok(allocated_eval)
+}
+
+// method compatible with src/spartan/sumcheck.rs > UniPoly::evaluate()
+pub fn uni_evaluate<Scalar, CS>(
+  mut cs: CS,
+  coeffs: Vec<AllocatedNum<Scalar>>,
+  r: AllocatedNum<Scalar>,
+) -> Result<AllocatedNum<Scalar>, SynthesisError>
+where
+  Scalar: PrimeField + PrimeFieldBits,
+  CS: ConstraintSystem<Scalar>,
+{
+  let mut eval: Num<Scalar> = Num::<Scalar>::from(coeffs[0].clone());
+  let mut power = r.clone();
+  for (i, coeff) in coeffs.iter().enumerate().skip(1) {
```

Comment:
These functions get quite complicated even though they are supposed to be doing simple things. This might just be how bellman is, or maybe you can define two helper methods here?

For example one could be `multiply_and_add()` and the other would be `multiply_with_r()` and both of them would return their `aux` variable? This way you could simplify this func, and you could also separately test the helper functions as well?